File: api/stream.js
import { getSession, invalidateSession } from './auth.js';

const VIDIO_API_KEY = 'CH1ZFsN4N/MIfAds1DL9mP151CNqIpWHqZGRr+LkvUyiq3FRPuP1Kt6aK+pG3nEC1FXt0ZAAJ5FKP8QU8CZ5/oOnzLxNND10Rm+tet/uqMtlFPMTc2yhz4IYyoRR1EiOiEIC3MckBPp+w1/0sDXRy+Ksnm1sOA+i8CqmbiBWOlU=';
const UA = 'Mozilla/5.0 (Linux; Android 15; Infinix X6880) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/148.0.7778.215 Mobile Safari/537.36';

async function fetchStream(channelId, cookie) {
  return fetch(
    `https://api.vidio.com/livestreamings/${channelId}/stream?initialize=true`,
    {
      headers: {
        'Host': 'api.vidio.com',
        'X-Api-Key': VIDIO_API_KEY,
        'X-API-Platform': 'web-mobile',
        'X-Secure-Level': '2',
        'Origin': 'https://www.vidio.com',
        'Referer': 'https://www.vidio.com/',
        'Accept': 'application/vnd.api+json',
        'Accept-Language': 'id-ID,id;q=0.9',
        'Cache-Control': 'no-cache',
        'Pragma': 'no-cache',
        'Cookie': cookie,
        'User-Agent': UA,
      }
    }
  );
}

export default async function handler(req, res) {
  const channelId = req.query.id || '875';

  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  try {
    // Ambil session (login jika belum / expired)
    let cookie = await getSession();
    let response = await fetchStream(channelId, cookie);

    // Jika 401/403, session mungkin expired → invalidate & retry sekali
    if (response.status === 401 || response.status === 403) {
      console.warn(`[stream] Session rejected (${response.status}), re-logging in...`);
      invalidateSession();
      cookie = await getSession();
      response = await fetchStream(channelId, cookie);
    }

    if (!response.ok) {
      const body = await response.text();
      return res.status(response.status).json({
        error: 'Gagal ambil stream dari Vidio',
        status: response.status,
        detail: body.slice(0, 500),
      });
    }

    const data = await response.json();

    // Coba berbagai path HLS yang mungkin ada di response
    const hlsUrl =
      data?.data?.attributes?.hls ||
      data?.data?.attributes?.hls_url ||
      data?.livestreaming?.hls ||
      data?.hls;

    if (!hlsUrl) {
      return res.status(404).json({
        error: 'HLS URL tidak ditemukan di response',
        available_keys: data?.data?.attributes ? Object.keys(data.data.attributes) : [],
      });
    }

    // Redirect ke HLS URL
    res.setHeader('Cache-Control', 'no-store');
    return res.redirect(302, hlsUrl);

  } catch (err) {
    console.error('[stream] Error:', err.message);
    return res.status(500).json({ error: err.message });
  }
}

File: api/stream.py
import json
import requests
from http.server import BaseHTTPRequestHandler

HEADERS = {
    "Accept": "*/*",
    "Accept-Encoding": "gzip",
    "Origin": "https://www.vidio.com",
    "Referer": "https://www.vidio.com/",
    "User-Agent": "Vidio/6.43.9-8ec34856ef (Linux;Android 11) ExoPlayerLib/2.19.1",
}


def get_stream_url(stream_id, stream_type="livestreamings"):
    url = f"https://api.vidio.com/{stream_type}/{stream_id}/stream?initialize=true"

    try:
        resp = requests.get(url, headers=HEADERS, timeout=10)
        resp.raise_for_status()
        data = resp.json()

        attrs = data.get("data", {}).get("attributes", {})
        hls = attrs.get("hls")
        if hls:
            return hls, None

        return None, f"HLS not found: {json.dumps(data)[:500]}"

    except requests.exceptions.HTTPError as e:
        return None, f"HTTP {e.response.status_code}: {e.response.text[:500]}"
    except Exception as e:
        return None, str(e)


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        from urllib.parse import urlparse, parse_qs

        parsed = urlparse(self.path)
        params = parse_qs(parsed.query)

        stream_id = params.get("id", ["204"])[0]
        stream_type = params.get("type", ["livestreamings"])[0]

        if stream_type not in ["livestreamings", "videos"]:
            self._error(400, "Invalid type")
            return

        hls_url, error = get_stream_url(stream_id, stream_type)

        if error:
            self._error(502, error)
            return

        self.send_response(302)
        self.send_header("Location", hls_url)
        self.send_header("Cache-Control", "no-cache")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()

    def _error(self, code, message):
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(json.dumps({"error": message}).encode())

    def log_message(self, format, *args):
        pass
